icu4c/source/i18n/unicode/messageformat2.h

```diff
@@ -0,0 +1,138 @@
+ *
+ * <p>See <a target="github" href="https://github.com/unicode-org/message-format-wg/blob/main/spec/syntax.md">the
+ * description of the syntax with examples and use cases</a> and the corresponding
+ * <a target="github" href="https://github.com/unicode-org/message-format-wg/blob/main/spec/message.abnf">ABNF</a> grammar.</p>
+ *
+ * @internal ICU 74.0 technology preview
+ * @deprecated This API is for technology preview only.
+ */
+
+class U_I18N_API MessageFormat2 : public Format {
+  public:
+    /**
+     * Constructs a new MessageFormat2 using the given pattern and the
+     * default locale.
+     *
+     * @param pattern   Pattern used to construct object.
+     * @param parseError Struct to receive information on the position
+     *                   of an error within the pattern.
+     * @param status    Input/output error code.  If the
+     *                  pattern cannot be parsed, set to failure code.
+     *
+     * @internal ICU 74.0 technology preview
+     * @deprecated This API is for technology preview only.
+     */
+    MessageFormat2(const UnicodeString &pattern, UParseError &parseError, UErrorCode &status);
+
+    /**
+     * Destructor.
+     *
+     * @internal ICU 74.0 technology preview
+     * @deprecated This API is for technology preview only.
+     */
+    virtual ~MessageFormat2();
+
+    // Not yet implemented
+    virtual bool operator==(const Format &other) const override;
+    // Not yet implemented
+    virtual bool operator!=(const Format &other) const;
+
+    // Not yet implemented
+    virtual UnicodeString &format(const Formattable &obj, UnicodeString &appendTo, FieldPosition &pos,
+                                  UErrorCode &status) const override;
+
+    // Not yet implemented
+    virtual void parseObject(const UnicodeString &source, Formattable &result,
+                             ParsePosition &pos) const override;
+
+    /**
+     * Clones this Format object polymorphically.  The caller owns the
+     * result and should delete it when done.
+     *
+     * @internal ICU 74.0 technology preview
+     * @deprecated This API is for technology preview only.
+     */
+
+    virtual MessageFormat2 *clone() const override;
+
+    /**
+     * Constructs a new MessageFormat2 from an existing one.
+     *
+     * @internal ICU 74.0 technology preview
+     * @deprecated This API is for technology preview only.
+     */
+    MessageFormat2(const MessageFormat2 &);
+
+    /**
+     * Returns a unique class ID POLYMORPHICALLY.  Pure virtual override.
+     * This method is to implement a simple version of RTTI, since not all
+     * C++ compilers support genuine RTTI.  Polymorphic operator==() and
+     * clone() methods call this method.
+     *
+     *
+     * @internal ICU 74.0 technology preview
+     * @deprecated This API is for technology preview only.
+     */
+    virtual UClassID getDynamicClassID(void) const override;
```

Comment:
I was getting this from the [coding guidelines](https://unicode-org.github.io/icu/userguide/dev/codingguidelines.html#c-class-layout-and-boilerplate): "If a class is a subclass of a parent (e.g., Format) with ICU’s “poor man’s RTTI” (Run-Time Type Information) mechanism (via getDynamicClassID() and getStaticClassID()) then add that to the new subclass as well (copy implementations from existing C++ APIs)." However, if `MessageFormat2` doesn't need to inherit from `Format`, then this can be removed too.